newa/__init__.py

```diff
@@ -129,6 +148,26 @@ class Event(Serializable):
     id: 'ErratumId'
 
 
+@frozen
+class ErrataTool:
+    """
+    Interface to Errata Tool instance
+
+    Its only purpose is to serve information about an erratum and its builds.
+    """
+
+    # TODO: Until we have a dedicated newa config, we'll setup ET instance
+    # url via environment variable ET_URL
+    url: str = field(default=os.environ.get("ET_URL"))
```

Comment:
Despite it is a temporary solution, we can expect users would forget to do such a setup. I would propose to fail with a reasonable errro when ET_URL is not set.

Reply:
I added field validator that would expect ET_URL to match a specific regexp:

```
    url: str = field(default=os.environ.get("ET_URL", default=""),
                     validator=validators.matches_re("^https?://.+$"))
```

If ET_URL is missing or seems to be invalid `ValueError` exception will pop up:

```
❯ REQUESTS_CA_BUNDLE=/etc/pki/tls/cert.pem newa  event -e 128049
Traceback (most recent call last):
  File "/home/omoris/.local/share/hatch/env/virtual/newa/6kDqWrPU/dev/bin/newa", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/omoris/.local/share/hatch/env/virtual/newa/6kDqWrPU/dev/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/omoris/.local/share/hatch/env/virtual/newa/6kDqWrPU/dev/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/omoris/.local/share/hatch/env/virtual/newa/6kDqWrPU/dev/lib/python3.11/site-packages/click/core.py", line 1719, in invoke
    rv.append(sub_ctx.command.invoke(sub_ctx))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/omoris/.local/share/hatch/env/virtual/newa/6kDqWrPU/dev/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/omoris/.local/share/hatch/env/virtual/newa/6kDqWrPU/dev/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/omoris/.local/share/hatch/env/virtual/newa/6kDqWrPU/dev/lib/python3.11/site-packages/click/decorators.py", line 45, in new_func
    return f(get_current_context().obj, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/omoris/Work/Projects/newa-omoris/newa/cli.py", line 111, in cmd_event
    errata = Erratum.from_errata_tool(event)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/omoris/Work/Projects/newa-omoris/newa/__init__.py", line 233, in from_errata_tool
    releases_json = ErrataTool().fetch_releases(event.id)
                    ^^^^^^^^^^^^
  File "<attrs generated init newa.ErrataTool>", line 5, in __init__
  File "/home/omoris/.local/share/hatch/env/virtual/newa/6kDqWrPU/dev/lib/python3.11/site-packages/attr/validators.py", line 146, in __call__
    raise ValueError(
ValueError: ("'url' must match regex '^https?://.+$' ('' doesn't)", Attribute(name='url', default='', validator=<matches_re validator for pattern re.compile('^https?://.+$')>, repr=True, eq=True, eq_key=None, order=True, order_key=None, hash=None, init=True, metadata=mappingproxy({}), type='str', converter=None, kw_only=False, inherited=False, on_setattr=None, alias='url'), re.compile('^https?://.+$'), '')
```

Will that do the trick? Or do you prefer something more specific (ie. an exception that directly mentions ET_URL)?

Reply:
I'd say the exception is correct: when there's no value provided, the default is used, and after that, the validator is applied, and `""` does not match the regular exception.

I'd suggest the following:

```python
     # TODO: Until we have a dedicated newa config, we'll setup ET instance
     # url via environment variable ET_URL
     def _url_factory() -> str:
         if 'ET_URL' in os.environ:
             return os.environ['ET_URL']

         raise Exception("ET_URL envvar is required.")

     url: str = field(factory=_url_factory,
                      validator=validators.matches_re("^https?://.+$"))
```

This way, we can tell user the `ET_URL` is required, validator will take care of validation, and we avoid the `""` which would be unusable anyway.

One more note: I'd recommend adding `NEWA_` prefix to the envvar name. Eventually, newa will be integrated with other tools, and it's very easy to add envvars for all those tools, it's also extremely hard to resolve any conflict should one of the tools claim `ET_URL` for its own goals, with a different meaning. By namespacing it, we can avoid any conflict with `ET_URL` required by TF or tmt (although they would consume `TESTING_FARM_ET_URL` and `TMT_ET_URL`, respectively, but you get the point, right? :)

Reply:
Agreed. Code updated (`ET_URL` -> `NEWA_ET_URL`) and added default as suggested (I used `@url.default` decorator instead of factory, it should make no difference).